### PR TITLE
FIX X-LoRA adapter lifecycle: disable_adapter, training mode and frozen classifier

### LINKED ISSUE #3635: X-LoRA: disable_adapter does not disable the experts, and generate() freezes the classifier / switches the model to eval mode
### System Info

- peft: `0.20.1.dev0` (main, `15c0feda`)
- transformers: 5.16.1, torch 2.13.0, Python 3.11, macOS, CPU

### Who can help?

@BenjaminBossan @githubnemo

### Reproduction

Four related bugs in X-LoRA's adapter lifecycle. All are reproduced with the tiny model used by `tests/test_xlora.py`.

```python
import os, tempfile, torch
from transformers import AutoModelForCausalLM
from peft import LoraConfig, XLoraConfig, get_peft_model

mid = "peft-internal-testing/tiny-random-OPTForCausalLM"
tmp, files = tempfile.mkdtemp(), []
for i in range(2):
    m = AutoModelForCausalLM.from_pretrained(mid)
    pm = get_peft_model(m, LoraConfig(task_type="CAUSAL_LM", target_modules=["q_proj", "v_proj"],
                                      init_lora_weights=False))
    d = os.path.join(tmp, f"a{i}"); pm.save_pretrained(d); files.append(d)

base = AutoModelForCausalLM.from_pretrained(mid); base.config.use_cache = False
cfg = XLoraConfig(task_type="CAUSAL_LM", hidden_size=base.config.hidden_size, xlora_depth=2,
                  adapters={str(i): f for i, f in enumerate(files)})
model = get_peft_model(base, cfg).eval()
ids = torch.randint(0, 100, (1, 6))

raw = AutoModelForCausalLM.from_pretrained(mid); raw.config.use_cache = False; raw.eval()
with torch.no_grad():
    ref = raw(input_ids=ids).logits
    on = model(input_ids=ids).logits
    with model.disable_adapter():
        off = model(input_ids=ids).logits
print("1) enabled vs base :", (on - ref).abs().max().item())
print("1) disabled vs base:", (off - ref).abs().max().item())   # expected 0.0

print("2) experts trainable after one forward:",
      sum(p.requires_grad for n, p in model.named_parameters() if "lora_A" in n or "lora_B" in n))

model.train()
model.generate(input_ids=ids[:, :4], max_new_tokens=4)
print("3) inner model training mode after generate:", model.base_model.lora_model.model.training)
print("4) classifier params still trainable:",
      sum(p.requires_grad for n, p in model.named_parameters() if "internal_xlora_classifier" in n))
```

Output on `main`:

```text
1) enabled vs base : 0.4138830602169037
1) disabled vs base: 0.5074496269226074
2) experts trainable after one forward: 40
3) inner model training mode after generate: False
4) classifier params still trainable: 0
```

**1. `disable_adapter()` does not disable X-LoRA.** `XLoraLinearLayer.forward` (and the Embedding/Conv2d variants) replace the wrapped LoRA layer's `forward` and never look at the disabled state. `XLoraModel.disable_adapter_layers()` only sets `self.disabled = True`, which stops `_enable_peft_forward_hooks` from injecting scalings — so inside the context every expert is applied with `scaling_weight = 1` instead of not being applied. Disabling the adapter therefore moves the output *further* from the base model. `tests/test_xlora.py::test_disable_adapter` only asserts that the outputs differ, so it passes.

**2. Frozen experts are silently unfrozen by a forward pass.** The scalings pass calls `self.lora_model.disable_adapter_layers()` and then `enable_adapter_layers()` in a `finally`; the latter goes through `set_adapter`, which sets `requires_grad=True` on all experts. `_maybe_freeze_all_adapters()` is only called after `generate()`, so a plain `forward` leaves all experts trainable even with the default `use_trainable_adapters=False`.

**3. `generate()` permanently switches the model into eval mode.** `_maybe_freeze_all_adapters()` calls `self.eval()`, and `generate()` calls it on every invocation. Generating once during training (evaluation, RL-style rollouts) disables dropout for the rest of the run. `PeftModel.training` still reports `True`, so this is invisible.

**4. `generate()` freezes the X-LoRA classifier.** `_maybe_freeze_all_adapters()` matches parameters with `"lora_" in name` while iterating over the whole `XLoraModel`. The classifier parameters are named `internal_xlora_classifier.layers.*`, and `"internal_xlora_classifier"` contains `"lora_"`. At construction time the classifier does not exist yet, so it survives; on the call after `generate()` it is frozen. Since the classifier is the only trainable part of X-LoRA when the experts are frozen, training after a `generate()` call is a silent no-op.

Additionally, the Hub-loading branch of `PeftModel.from_pretrained` builds the expert paths as `os.path.join(model_id, model_id)` (`peft_model.py:575`), which is not a valid repo id — the subfolder is passed separately. That path is currently unreachable anyway because of the unpacking bug on `xlora/model.py:270`, which is already fixed in #3421.

### Expected behavior

- `disable_adapter()` reproduces the base model output.
- Experts stay frozen unless `use_trainable_adapters=True`, regardless of how many forward passes were run.
- `generate()` does not change the model's training mode.
- The X-LoRA classifier stays trainable.

### Would a PR be welcome?

I have a fix ready with five regression tests that fail on `main` and pass with the change (`pytest tests/test_xlora.py`: 21 passed; on `main` the five new tests fail). Per the contribution guidelines I'll wait for a maintainer go-ahead before opening it.

### AI assistance disclosure

AI assistance was used to find and prepare this. I have reviewed every line and run the reproductions and the test suite myself.

### PR DESCRIPTION
**Pull request by amogh-nagri-11**
Fixes #3635.

## What this fixes

Four related bugs in X-LoRA's adapter lifecycle, all reproduced in #3635:

1. **`disable_adapter()` did not disable the experts.** The `XLoraLayer` forwards replace the wrapped LoRA layer's `forward` and never looked at the disabled state. Inside a `disable_adapter` context no scalings are injected, so every expert was applied with `scaling_weight = 1` instead of not being applied — disabling moved the output *further* from the base model. The three layer forwards (`Linear`, `Embedding`, `Conv2d`) now return the base layer output when the X-LoRA adapter is disabled.

2. **A plain forward pass unfroze the experts.** The scalings pass re-enables the LoRA layers in a `finally`, and `enable_adapter_layers()` goes through `set_adapter`, which sets `requires_grad=True`. `_maybe_freeze_all_adapters()` only ran after `generate()`, so a `forward` left all experts trainable even with the default `use_trainable_adapters=False`. It is now called right after `enable_adapter_layers()`.

3. **`generate()` switched the model to eval mode permanently.** `_maybe_freeze_all_adapters()` called `self.eval()`. Generating once during training silently disabled dropout for the rest of the run, and `PeftModel.training` still reported `True`. The `self.eval()` call moved to the end of `XLoraModel.__init__`, so construction behaviour is unchanged but `generate()` no longer has this side effect.

4. **`generate()` froze the X-LoRA classifier.** `_maybe_freeze_all_adapters()` matched `"lora_" in name` while iterating over the whole `XLoraModel`, and the classifier parameters are named `internal_xlora_classifier.layers.*` — which contains `"lora_"`. The classifier survives the call in `__init__` only because it does not exist yet at that point. Since it is the only trainable part of X-LoRA when the experts are frozen, training after a `generate()` call was a silent no-op. The loop now iterates over `self.lora_model`, which excludes the classifier.

Also fixes the X-LoRA branch of `PeftModel.from_pretrained`, which built the expert paths as `os.path.join(model_id, model_id)` — not a valid Hub repo id, the subfolder is passed separately. That branch is currently unreachable because of the unpacking bug on `xlora/model.py`, which #3421 fixes; this one-line change is reasoned and reviewed but not covered by a test, since exercising it needs a Hub repo with per-adapter subfolders.

## Tests

Five regression tests added to `tests/test_xlora.py`, each failing on `main` and passing with this change:

- `test_disable_adapter_matches_base_model` and `test_disable_adapter_matches_base_model_embedding` — the disabled output equals the base model output (the existing `test_disable_adapter` only asserts that the outputs differ, so it passed with the bug).
- `test_generate_preserves_training_mode`
- `test_classifier_stays_trainable_after_generate`
- `test_experts_stay_frozen_after_forward`

Commands run (macOS, CPU, transformers 5.16.1, torch 2.13.0, Python 3.11):

```
pytest tests/test_xlora.py -q          # 21 passed
pytest tests/test_xlora.py -q -k "disable_adapter_matches or preserves_training or stays_trainable or stay_frozen"
                                       # 5 failed on main, 5 passed with this change
make style                             # clean
```

I also ran the wider suite (`pytest tests/ --ignore=tests/test_gpu_examples.py --ignore=tests/test_common_gpu.py --ignore=tests/regression`) before and after the change: no new failures. The only failures on this machine are 12 pre-existing `test_initialization.py::*conversion_same_output_after_loading*` cases, which miss the hard-coded `1e-6` tolerance by ~1.7e-6 in float32 on CPU here and are unrelated to this change.

No GPU was available, so the GPU test files were not run.

## AI assistance disclosure

AI assistance was used to find these bugs and prepare the patch. I have reviewed every changed line, ran the reproductions and the test commands above myself, and can defend the change end to end.

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -572,7 +572,7 @@ def from_pretrained(
                     # Prepare a dict of adapter paths, which really just point to the hf id; we will use the subfolders
                     adapter_paths = {}
                     for loaded_adapter_name in adapter_names:
-                        adapter_paths[loaded_adapter_name] = os.path.join(model_id, model_id)
+                        adapter_paths[loaded_adapter_name] = model_id
```
Comment (BenjaminBossan) [inline src/peft/peft_model.py]: This is not tested anywhere, is it?

File: tests/test_xlora.py [hunk 2]
```diff
@@ -349,6 +349,82 @@ def test_disable_adapter(self, tokenizer, model):
...
+    def test_generate_preserves_training_mode(self, tokenizer, model):
+        # generate used to put the whole model into eval mode as a side effect, which silently disables dropout for
+        # the rest of the training run
+        inputs = tokenizer.encode("Python is a", add_special_tokens=False, return_tensors="pt")
+        model.train()
+
+        model.generate(input_ids=inputs.to(self.torch_device), max_new_tokens=4)
+
+        assert model.base_model.training
+        assert model.base_model.lora_model.model.training
```
Comment (BenjaminBossan) [inline tests/test_xlora.py]: Let's also test that if the model is in `eval` mod, it's preserved too.

File: src/peft/tuners/xlora/model.py [hunk 3]
```diff
@@ -307,10 +307,14 @@ def __init__(
...
-    def _maybe_freeze_all_adapters(self):
         self.eval()
+
+    def _maybe_freeze_all_adapters(self):
         if not self.xlora_config.use_trainable_adapters:
-            for name, param in self.named_parameters():
+            # Only the LoRA experts are frozen. Iterating over the LoRA model instead of over self excludes the
+            # X-LoRA classifier, whose parameter names also contain "lora_" (via "internal_xlora_classifier") and
+            # which is the only trainable part of X-LoRA when the experts are frozen.
+            for name, param in self.lora_model.named_parameters():
```
Comment (BenjaminBossan) [inline src/peft/tuners/xlora/model.py]: What about checking for `".lora_"` instead of `"lora_"`? That should correctly identify the LoRA layers but exclude `"xlora_"`.

File: tests/test_xlora.py [hunk 4]
```diff
@@ -349,6 +349,82 @@ def test_disable_adapter(self, tokenizer, model):
...
+        # substring that is used to identify the frozen experts. The classifier is the only trainable part of X-LoRA
+        # when use_trainable_adapters=False, so freezing it makes training a silent no-op.
+        classifier_params = [param for name, param in model.named_parameters() if "internal_xlora_classifier" in name]
+        assert classifier_params
+        assert all(param.requires_grad for param in classifier_params)
+
+        inputs = tokenizer.encode("Python is a", add_special_tokens=False, return_tensors="pt")
+        model.generate(input_ids=inputs.to(self.torch_device), max_new_tokens=4)
+
+        assert all(param.requires_grad for param in classifier_params)
```
Comment (BenjaminBossan) [inline tests/test_xlora.py:421]: Let's do the sanity check that the other parameters don't require grads.

File: tests/test_xlora.py [hunk 5]
```diff
@@ -349,6 +349,82 @@ def test_disable_adapter(self, tokenizer, model):
...
+
+    def test_experts_stay_frozen_after_forward(self, tokenizer, model):
+        # The scalings pass disables and re-enables the LoRA layers, and enabling them marks them as trainable again.
```
Comment (BenjaminBossan) [inline tests/test_xlora.py]: Let's make it crystal clear what we actually expect here.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for providing the fixes and clear tests. I have a couple of smaller issues, but generally the changes look correct.

File: src/peft/peft_model.py (hunk 1 above)
Comment (amogh-nagri-11) [reply src/peft/peft_model.py]: Correct, it is not. That branch only runs when `PeftModel.from_pretrained` is handed a model whose `base_model` is already an `XLoraModel` **and** `model_id` is a Hub repo with one subfolder per expert, and there is no such test repo — covering it would have meant mocking `HfFileSystem`, `PeftConfig._get_peft_type` and the weight download, which mostly tests the mocks.

I have dropped the hunk, so this PR is now only what the tests cover. The line is still wrong (`os.path.join("org/repo", "org/repo")` is not a valid repo id), but it is unreachable anyway until the unpacking bug on `xlora/model.py` is fixed by #3421, so it seems better to fold it into that work than to carry an untested change here. It is written up in point 3 of #3635 so it does not get lost.

File: src/peft/tuners/xlora/model.py (hunk 3 above)
Comment (amogh-nagri-11) [reply src/peft/tuners/xlora/model.py]: Good call, that is cleaner than restricting the iteration. Done.

I checked the two directions on the `model` fixture: `".lora_"` matches all 40 expert parameters (`lora_model.model.…self_attn.v_proj.lora_A.0.weight`), none of the 4 classifier parameters (`internal_xlora_classifier.layers.0.weight`), and nothing else. It also correctly does not match the `lora_model.` prefix itself, since that has no leading dot.

File: tests/test_xlora.py (hunk 2 above)
Comment (amogh-nagri-11) [reply tests/test_xlora.py]: Done — the test is now parametrized over `training` and asserts the mode in both directions.

Note that the `False` case passes on `main` too (the old code called `self.eval()`, so eval mode happened to survive); it is there as a guard rather than as a regression test.

File: tests/test_xlora.py (hunk 4 above)
Comment (amogh-nagri-11) [reply tests/test_xlora.py:421]: Done. The test now also collects the non-classifier parameters and asserts that none of them require grads, both before and after `generate`, so it states the actual invariant: with `use_trainable_adapters=False` the classifier is the only trainable part of the model.

File: tests/test_xlora.py (hunk 5 above)
Comment (amogh-nagri-11) [reply tests/test_xlora.py]: Done. The comment now says what is expected rather than only what went wrong: with `use_trainable_adapters=False` the LoRA experts must never require grads regardless of how many forward passes were run, and it explains why a plain forward used to unfreeze them (the scalings pass re-enables the LoRA layers through `set_adapter`, and the experts were only frozen again after `generate`).

I also tightened the parameter filter from `"lora_A" in name` to `".lora_A." in name` to match the fix.

**Comment by amogh-nagri-11** [comment]
Thanks for the review. All five points are addressed in be387792 — replies inline.

Summary:
- `_maybe_freeze_all_adapters` now matches `".lora_"` on `self.named_parameters()`, as suggested, rather than restricting the iteration.
- `test_generate_preserves_training_mode` is parametrized over train/eval.
- `test_classifier_stays_trainable_after_generate` now asserts that the classifier is the *only* trainable part, before and after `generate`.
- `test_experts_stay_frozen_after_forward` spells out the expected invariant.
- The untested `peft_model.py` one-liner is dropped; it is unreachable until #3421 lands and is recorded in #3635 instead.

`pytest tests/test_xlora.py` → 22 passed. Against `main`, 5 of the 6 new cases fail (`test_generate_preserves_training_mode[False]` passes there, it is a guard rather than a regression test). `make style` is clean.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for the X-LoRA fixes, they LGTM.
